### PR TITLE
HHH-15328 Add support for CTE WITH clause

### PR DESCRIPTION
**Pull request by beikov**
https://hibernate.atlassian.net/browse/HHH-15328

This is a draft at this point. I want to see tests running before splitting this PR up into smaller pieces.

### CODE REVIEW

**Comment by gavinking** [comment]
Awesome, I was wondering about this over the w/e!

File: hibernate-core/src/main/java/org/hibernate/query/criteria/JpaCteContainer.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+	/**
+	 * Allows to register a recursive CTE. The base {@link CriteriaQuery} serves
+	 * for the structure of the {@link JpaCteCriteria}, which is made available in the recursive criteria producer function,
+	 * so that the recursive {@link CriteriaQuery} is able to refer to the CTE again.
+	 *
+	 * @see JpaCriteriaQuery#from(JpaCteCriteria)
+	 * @see JpaFrom#join(JpaCteCriteria, SqmJoinType)
+	 */
+	<T> JpaCteCriteria<T> with(AbstractQuery<T> baseCriteria, boolean unionDistinct, Function<JpaCteCriteria<T>, AbstractQuery<T>> recursiveCriteriaProducer);
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/criteria/JpaCteContainer.java:49]: Let's call it `withRecursiveUnion` and `withRecursiveUnionAll` to get rid of the boolean flag.

File: hibernate-core/src/main/java/org/hibernate/query/criteria/JpaCteCriteria.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+	default void cycle(String cycleMarkAttributeName, String cyclePathAttributeName, JpaCteCriteriaAttribute... cycleColumns) {
+		cycle( cycleMarkAttributeName, cyclePathAttributeName, Arrays.asList( cycleColumns ) );
+	}
+
+	default void cycle(String cycleMarkAttributeName, String cyclePathAttributeName, List<JpaCteCriteriaAttribute> cycleColumns) {
+		cycle( cycleMarkAttributeName, cyclePathAttributeName, true, false, cycleColumns );
+	}
+
+	default <X> void cycle(String cycleMarkAttributeName, String cyclePathAttributeName, X cycleValue, X noCycleValue, JpaCteCriteriaAttribute... cycleColumns) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/criteria/JpaCteCriteria.java:102]: Rename the variant with the `path` to `cycleUsing`

**Comment by gavinking** [comment]
👏👏👏

**Pull request merged**